### PR TITLE
SONARPY-1461 S4507: Add support for Flask applications.

### PR DESCRIPTION
**Pull request by jdodinh**
This is a working implementation. However, I believe that the manner in which we raise the issues could be improved. I'm not sure, however, whether doing this in the context of this ticket is necessarily worth it.

### CODE REVIEW

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java [hunk 1]
```diff
@@ -21,52 +21,111 @@
...
+        .map(Symbol::fullyQualifiedName)
+        .filter(FLASK_APP_CONFIG_FQN::equals)
+        .isPresent() && Optional.of(subscriptionExpression.subscripts())
+        .map(ExpressionList::expressions)
+        .filter(list -> list.size() == 1)
+        .map(list -> list.get(0))
+        .flatMap(TreeUtils.toOptionalInstanceOfMapper(StringLiteral.class))
+        .map(StringLiteral::trimmedQuotesValue)
+        .filter("DEBUG"::equals)
+        .isPresent();
```
Comment (joke1196) [inline python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java:126]: Maybe here we could refactor the two parts of the boolean into functions just to make it clearer about the intent. i.e. `isFlaskAppConfig` && `hasDebugModeEnabled` or something like that.

File: python-checks/src/test/resources/checks/hotspots/debugMode/settings.py [hunk 2]
```diff
@@ -2,4 +2,12 @@
...
 DEBUG = False  # OK
+DEBUG = print()  # OK
+DEBUG_PROPAGATE_EXCEPTIONS = False  # OK
 Other = True
+
+class MyClass:
+    def __init__(self):
+        self.boolean_property = True
+my_object = MyClass()
+my_object.boolean_property = True
```
Comment (joke1196) [inline python-checks/src/test/resources/checks/hotspots/debugMode/settings.py:13]: Not sure exactly what is tested here, wouldn't this be covered by the `Other` case already?

**Review by joke1196 — CHANGES_REQUESTED** [review]
Looks good! Good job on the tests thinking about all those edge cases. I have added a few comments that may help with the readability of the code. There is also something I didn't understand in the tests for the `settings.py` file

File: python-checks/src/test/resources/checks/hotspots/debugMode/settings.py (hunk 2 above)
Comment (jdodinh) [reply python-checks/src/test/resources/checks/hotspots/debugMode/settings.py:13]: This is just to increase the coverage. It is important to have an fqn which corresponds to a boolean property on the lhs. This is to fix the low coverage that occurred on the existing code. 

Since we want our test files to only have valid python code, I decided it would be simplest to create this class for the test.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java [hunk 3]
```diff
@@ -21,52 +21,111 @@
...
+      arguments.stream().filter(DebugModeCheck::isDebugArgument).forEach(arg -> ctx.addIssue(arg, MESSAGE));
+    }
+
+    if (FLASK_RUN_FQN.equals(getQualifiedName(callExpression))) {
+      RegularArgument debugArgument = TreeUtils.nthArgumentOrKeyword(2, "debug", arguments);
+      Optional.ofNullable(debugArgument)
+        .map(RegularArgument::expression)
+        .flatMap(TreeUtils.toOptionalInstanceOfMapper(Name.class))
+        .map(Name::name)
+        .filter(TRUE_KEYWORD::equals)
```
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java]: Yes, good catch. Made this change in 3478ad8373789abec94fd936a3bb673cc867c9f1.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java [hunk 4]
```diff
@@ -21,52 +21,111 @@
...
+        .ifPresent(str -> ctx.addIssue(debugArgument, MESSAGE));
+    }
+  }
+
+  private static void assignmentStatementConsumer(SubscriptionContext ctx) {
+    AssignmentStatement assignmentStatementTree = (AssignmentStatement) ctx.syntaxNode();
```
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java:81]: The reason is that we perform two different actions for the Assignment Statement node. For the Django FQNs, we need to check the file, but for the flask we don't. Therefore putting the return statement early implied a lot of code duplication. But I will try to think of a smarter way to go about this.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java [hunk 5]
```diff
@@ -21,52 +21,111 @@
...
+      if (isModifyingDebugProperty && isTrueLiteral(assignmentStatementTree.assignedValue())) {
+        ctx.addIssue(assignmentStatementTree, MESSAGE);
       }
-    });
+    }
+  }
+
+  private static boolean isGlobalDebugProperties(ExpressionList expressionList, SubscriptionContext ctx) {
+    return expressionList.expressions().stream().anyMatch(DebugModeCheck::isDebugIdentifier)
+      && settingFiles.contains(ctx.pythonFile().fileName());
```
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java:93]: I think that it is worth discussing this, as I am not sure what the best way would be otherwise.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java (hunk 1 above)
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java]: Made the change in b19304898ea7456a2094bfd4607017b0420f105d and further in a34fd1b446737eb707412e6fc2427cc559630607.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java:90]: Thanks for the chat. I just pushed new changes from c2e48c974482e84baf975ae3b788ae7ee9745877.

File: python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (jdodinh) [reply python-checks/src/main/java/org/sonar/python/checks/hotspots/DebugModeCheck.java]: Resolved in c2e48c974482e84baf975ae3b788ae7ee9745877.

**Review by joke1196 — APPROVED** [review]
LGTM! Thank you for the great work!
